src/coreclr/jit/valuenum.cpp

```diff
@@ -4917,8 +4917,10 @@ void Compiler::fgValueNumberLocalStore(GenTree*             storeNode,
                     // Avoid redundant bitcasts for the common case of a full definition.
                     fieldStoreType = fieldVarDsc->TypeGet();
                 }
+
+                ssize_t      fieldValueOffset = max(0, fieldVarDsc->lvFldOffset - offset);
```

Comment:
Not clear to me what this `max` is doing.

Reply:
The `max` chooses between the two possible cases:
```
       <---->      - "fieldValueOffset = fieldOffset - offset"
             [--]  - fieldStoreValue, also a full field local
 field offset:
       [---------]  - value
 offset:
[-----------------] - full local
```
```
                     - "fieldValueOffset = 0"
              [-|--] - fieldStoreValue, part of the field local
  field offset: 
                [-----]  - value
          offset:
[----------------------] - full local
```

Reply:
Perhaps a comment is warranted? 

I guess what confused me was that different aspects of non-entire writes are handled in different places, so it wasn't clear what part of the computation had to happen here.

Reply:
> Perhaps a comment is warranted?

Would this diff make it clearer?
```diff
-                ssize_t      fieldValueOffset = max(0, fieldVarDsc->lvFldOffset - offset);
+                // Calculate offset of this field's value, relative to the entire one.
+                ssize_t      fieldOffset      = fieldVarDsc->lvFldOffset;
+                ssize_t      fieldValueOffset = (fieldOffset < offset) ? 0 : (fieldOffset - offset);
```

> I guess what confused me was that different aspects of non-entire writes are handled in different places, so it wasn't clear what part of the computation had to happen here.

There are two parts to the computations: the "load" part, where we decide what value are we to store, it has as "the base" for the offsets "value", and then there is the "store" part (i. e. `processDef`), where "the base" is the field local. One can notice sizes of both the "load" and "store" parts are the same, so only the offsets need to be calculated differently.

Reply:
I think it helps, yeah. The other idea I had was to add this as yet another out arg from `gtStoreDefinesField` as it complements `pFieldStoreOffset` but I suppose there are consumers that don't' really care to know this information.